File: app/src/main/java/io/github/daylight00/nativeshell/TerminalSession.java
package io.github.daylight00.nativeshell;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

final class TerminalSession implements AutoCloseable {
    interface Listener {
        void onScreenChanged();

        void onSessionFinished(int exitCode, String errorMessage);
    }

    private static final int SIGNAL_HANGUP = 1;

    private final TerminalEmulator emulator;
    private final Listener listener;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final ExecutorService writer = Executors.newSingleThreadExecutor(runnable -> {
        Thread thread = new Thread(runnable, "native-shell-writer");
        thread.setDaemon(true);
        return thread;
    });
    private final AtomicBoolean closing = new AtomicBoolean();
    private final long handle;
    private final Thread readerThread;

    TerminalSession(Context context, int rows, int columns, Listener listener) throws IOException {
        this.listener = listener;
        emulator = new TerminalEmulator(rows, columns);
        String home = context.getFilesDir().getAbsolutePath();
        String temporaryDirectory = context.getCacheDir().getAbsolutePath();
        handle = NativePty.spawn(
                "/system/bin/sh",
                home,
                home,
                temporaryDirectory,
                rows,
                columns);
        if (handle == 0) {
            throw new IOException("native PTY returned an invalid handle");
        }
        readerThread = new Thread(this::readLoop, "native-shell-reader");
        readerThread.setDaemon(true);
        readerThread.start();
    }

    TerminalBuffer.Snapshot snapshot() {
        return emulator.snapshot();
    }

    void resize(int rows, int columns, int pixelWidth, int pixelHeight) {
        emulator.resize(rows, columns);
        try {
            NativePty.resize(handle, rows, columns, pixelWidth, pixelHeight);
        } catch (IOException error) {
            finish(-1, error.getMessage());
        }
    }

    void sendText(String text) {
        sendBytes(text.getBytes(StandardCharsets.UTF_8));
    }

    void sendByte(int value) {
        sendBytes(new byte[]{(byte) value});
    }

    void sendBytes(byte[] bytes) {
        if (bytes.length == 0 || closing.get()) {
            return;
        }
        byte[] copy = bytes.clone();
        writer.execute(() -> {
            try {
                NativePty.write(handle, copy, 0, copy.length);
            } catch (IOException error) {
                finish(-1, error.getMessage());
            }
        });
    }

    private void readLoop() {
        byte[] bytes = new byte[16 * 1024];
        int exitCode = -1;
        String errorMessage = null;
        try {
            while (!closing.get()) {
                int count = NativePty.read(handle, bytes, 0, bytes.length);
                if (count <= 0) {
                    break;
                }
                emulator.consume(bytes, 0, count);
                mainHandler.post(listener::onScreenChanged);
            }
            exitCode = NativePty.waitFor(handle);
        } catch (IOException error) {
            errorMessage = error.getMessage();
        } finally {
            closing.set(true);
            writer.shutdownNow();
            NativePty.destroy(handle);
            finish(exitCode, errorMessage);
        }
    }

    private void finish(int exitCode, String errorMessage) {
        mainHandler.post(() -> listener.onSessionFinished(exitCode, errorMessage));
    }

    @Override
    public void close() {
        if (!closing.compareAndSet(false, true)) {
            return;
        }
        writer.shutdownNow();
        try {
            NativePty.signalProcessGroup(handle, SIGNAL_HANGUP);
        } catch (IOException ignored) {
            // The child may already have exited.
        }
    }
}

File: app/src/main/kotlin/io/github/daylight00/nativeshell/TerminalSession.kt
package io.github.daylight00.nativeshell

import android.system.OsConstants
import java.io.File
import java.io.IOException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

internal class TerminalSession(
    private val homeDirectory: File,
    private val temporaryDirectory: File,
    private val listener: Listener,
) : AutoCloseable {
    interface Listener {
        fun onOutput(bytes: ByteArray)
        fun onExit(exitCode: Int)
        fun onFailure(error: Throwable)
    }

    private val closed = AtomicBoolean(false)
    private val nativeHandleLock = Any()
    private val commandExecutor: ExecutorService = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "native-shell-command").apply { isDaemon = true }
    }

    @Volatile
    private var handle: Long = 0L

    @Volatile
    private var pendingSize = Size(24, 80, 0, 0)

    fun start(rows: Int, columns: Int, pixelWidth: Int, pixelHeight: Int) {
        pendingSize = Size(rows, columns, pixelWidth, pixelHeight).sanitized()
        Thread({ runSession() }, "native-shell-reader").apply {
            isDaemon = true
            start()
        }
    }

    fun write(bytes: ByteArray) {
        if (bytes.isEmpty() || closed.get()) return
        val copy = bytes.copyOf()
        submitCommand {
            synchronized(nativeHandleLock) {
                val current = handle
                if (current == 0L || closed.get()) return@synchronized
                NativePty.write(current, copy, 0, copy.size)
            }
        }
    }

    fun resize(rows: Int, columns: Int, pixelWidth: Int, pixelHeight: Int) {
        val size = Size(rows, columns, pixelWidth, pixelHeight).sanitized()
        pendingSize = size
        submitCommand {
            synchronized(nativeHandleLock) {
                val current = handle
                if (current == 0L || closed.get()) return@synchronized
                NativePty.resize(
                    current,
                    size.rows,
                    size.columns,
                    size.pixelWidth,
                    size.pixelHeight,
                )
            }
        }
    }

    override fun close() {
        if (!closed.compareAndSet(false, true)) return
        try {
            commandExecutor.execute {
                synchronized(nativeHandleLock) {
                    val current = handle
                    if (current == 0L) return@synchronized
                    try {
                        NativePty.signalProcessGroup(current, OsConstants.SIGHUP)
                    } catch (_: IOException) {
                        try {
                            NativePty.signalProcessGroup(current, OsConstants.SIGKILL)
                        } catch (_: IOException) {
                            // The reader thread remains the native-session owner.
                        }
                    }
                }
            }
        } catch (_: RejectedExecutionException) {
            // Natural process exit can win the race with Activity destruction.
        } finally {
            commandExecutor.shutdown()
        }
    }

    private fun submitCommand(command: () -> Unit) {
        if (closed.get() || commandExecutor.isShutdown) return
        try {
            commandExecutor.execute {
                try {
                    command()
                } catch (error: IOException) {
                    if (!closed.get()) listener.onFailure(error)
                }
            }
        } catch (_: RejectedExecutionException) {
            // The reader thread has already completed and closed the command lane.
        }
    }

    private fun runSession() {
        var nativeHandle = 0L
        var exitCode: Int? = null
        try {
            val initial = pendingSize
            nativeHandle = NativePty.spawn(
                SHELL_PATH,
                homeDirectory.absolutePath,
                homeDirectory.absolutePath,
                temporaryDirectory.absolutePath,
                initial.rows,
                initial.columns,
            )
            synchronized(nativeHandleLock) {
                handle = nativeHandle
            }

            if (closed.get()) {
                try {
                    NativePty.signalProcessGroup(nativeHandle, OsConstants.SIGHUP)
                } catch (_: IOException) {
                    // waitFor below remains the authority for process completion.
                }
            } else {
                val latest = pendingSize
                NativePty.resize(
                    nativeHandle,
                    latest.rows,
                    latest.columns,
                    latest.pixelWidth,
                    latest.pixelHeight,
                )
            }

            val buffer = ByteArray(READ_BUFFER_SIZE)
            while (!closed.get()) {
                val count = NativePty.read(nativeHandle, buffer, 0, buffer.size)
                if (count <= 0) break
                listener.onOutput(buffer.copyOf(count))
            }

            exitCode = NativePty.waitFor(nativeHandle)
        } catch (error: Throwable) {
            if (!closed.get()) listener.onFailure(error)
        } finally {
            closed.set(true)
            synchronized(nativeHandleLock) {
                handle = 0L
            }
            commandExecutor.shutdownNow()
            val commandsStopped = try {
                commandExecutor.awaitTermination(COMMAND_STOP_TIMEOUT_SECONDS, TimeUnit.SECONDS)
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
                false
            }
            if (nativeHandle != 0L && commandsStopped) {
                NativePty.destroy(nativeHandle)
            } else if (nativeHandle != 0L) {
                listener.onFailure(
                    IllegalStateException("native command lane did not stop; session cleanup was deferred"),
                )
            }
        }
        if (exitCode != null) listener.onExit(exitCode)
    }

    private data class Size(
        val rows: Int,
        val columns: Int,
        val pixelWidth: Int,
        val pixelHeight: Int,
    ) {
        fun sanitized(): Size = Size(
            rows.coerceIn(1, MAX_ROWS),
            columns.coerceIn(1, MAX_COLUMNS),
            pixelWidth.coerceIn(0, MAX_PIXELS),
            pixelHeight.coerceIn(0, MAX_PIXELS),
        )
    }

    private companion object {
        const val SHELL_PATH = "/system/bin/sh"
        const val READ_BUFFER_SIZE = 16 * 1024
        const val MAX_ROWS = 2_000
        const val MAX_COLUMNS = 2_000
        const val MAX_PIXELS = 65_535
        const val COMMAND_STOP_TIMEOUT_SECONDS = 5L
    }
}
